[FSDP][1/N] Refactor module materialization

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #94198 [FSDP][3/N] Add LCA logic to `fully_shard`
* #94197 [FSDP][2/N] Add util for computing shared param LCA
* **#94196 [FSDP][1/N] Refactor module materialization**

**Overview**
This refactors module materialization (i.e. meta device or `torchdistX` deferred initialization) to compute the parameter and buffer names as needed instead of pre-computing them. These are needed to reacquire references to the states (e.g. `module.get_parameter(param_name)`) after materialization since the materialization may create new variables.

This refactor simplifies `_get_fully_sharded_module_to_states()` (the core function for "pseudo auto wrapping") to better enable lowest common ancestor (LCA) module computation for shared parameters, for which tracking parameter and buffer names may complicate the already non-obvious implementation.

**Discussion**
The tradeoff is a worst case quadratic traversal over modules if materializing all of them. However, since (1) the number of modules is relatively small, (2) the computation per module in the quadratic traversal is negligible, (3) this runs only once per training session, and (4) module materialization targets truly large models, I think this tradeoff is tolerable.

**For Reviewers**
- `_init_param_handle_from_module()` initializes _one_ `FlatParamHandle` from a fully sharded module and represents the module wrapper code path. For this code path, there is no need to reacquire references to the parameters/buffers for now since the managed parameters are only computed after materialization. This works because the managed parameters have a simple definition: any parameter in the local root module's tree excluding those already marked as flattened by FSDP. Similarly, FSDP marks buffers to indicate that they have already been processed (synced if `sync_module_states`).
- `_init_param_handles_from_module()` initializes _all_ `FlatParamHandle`s from a fully sharded module and represents the composable code path. For this code path, we must reacquire references to parameters/buffers because each logical wrapping is specified as a list of parameters/buffers to group together by those variables and because materialization may create new variables.

Commit:
[FSDP][1/N] Refactor module materialization

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -395,13 +395,19 @@ def _init_param_handle_from_module(
+        fully_sharded_module, state._ignored_params
     )
+    # Materialize the module if needed
+    if (is_meta_module or is_torchdistX_deferred_init) and param_init_fn is not None:
```
I thought if param_init_fn is None, but we have meta modules / params, we should try to do some default initialization for the user?

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -453,19 +459,33 @@ def _init_param_handles_from_module(
+        if is_meta_module or is_torchdistX_deferred_init:
+            materialized_module = True
+            # Save the parameter and buffer names to reacquire references after
+            # after materialization since their variables may change
```
Even after reading the PR description, I'm not 100% sure why the variables may change after materialization? I thought materialization is all about filling in meta parameters with their actual values?

awgu replied:
Yes, but at the implementation level, `module.to(device)` will replace all meta-device parameters with new Python parameter variables.

This function returns `False` for meta-device tensors:
https://github.com/pytorch/pytorch/blob/a064ce19391c5ac2d16f03a7b77735a70ff93a39/torch/nn/modules/module.py#L799-L800
This leads to the `else` branch:
https://github.com/pytorch/pytorch/blob/a064ce19391c5ac2d16f03a7b77735a70ff93a39/torch/nn/modules/module.py#L821-L829

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -453,19 +459,33 @@ def _init_param_handles_from_module(
+        elif is_meta_module:
+            _materialize_meta_module(fully_sharded_module, device_id)
+        elif is_torchdistX_deferred_init:
+            deferred_init.materialize_module(
```
do we have unittests covering deferred init for the composable path?

awgu replied:
We do not because torchdistX does not support the latest PyTorch version, so I could not run torchdistX locally.

rohan-varma reviewed (approved):
LGTM thanks!

Commit:
Update on "[FSDP][1/N] Refactor module materialization"


**Overview**
This refactors module materialization (i.e. meta device or `torchdistX` deferred initialization) to compute the parameter and buffer names as needed instead of pre-computing them. These are needed to reacquire references to the states (e.g. `module.get_parameter(param_name)`) after materialization since the materialization may create new variables.

This refactor simplifies `_get_fully_sharded_module_to_states()` (the core function for "pseudo auto wrapping") to better enable lowest common ancestor (LCA) module computation for shared parameters, for which tracking parameter and buffer names may complicate the already non-obvious implementation.

**Discussion**
The tradeoff is a worst case quadratic traversal over modules if materializing all of them. However, since (1) the number of modules is relatively small, (2) the computation per module in the quadratic traversal is negligible, (3) this runs only once per training session, and (4) module materialization targets truly large models, I think this tradeoff is tolerable.

**For Reviewers**
- `_init_param_handle_from_module()` initializes _one_ `FlatParamHandle` from a fully sharded module and represents the module wrapper code path. For this code path, there is no need to reacquire references to the parameters/buffers for now since the managed parameters are only computed after materialization. This works because the managed parameters have a simple definition: any parameter in the local root module's tree excluding those already marked as flattened by FSDP. Similarly, FSDP marks buffers to indicate that they have already been processed (synced if `sync_module_states`).
- `_init_param_handles_from_module()` initializes _all_ `FlatParamHandle`s from a fully sharded module and represents the composable code path. For this code path, we must reacquire references to parameters/buffers because each logical wrapping is specified as a list of parameters/buffers to group together by those variables and because materialization may create new variables.


[ghstack-poisoned]

awgu commented:
@pytorchbot merge
